### PR TITLE
Refactor code to improve unread marker logic

### PR DESCRIPTION
**Pull request by gedu**
### Details

- Migrates Unread Marker logic from ReportActionsView to ReportActionsList component
- Improves the unread marker logic by using hooks

### Fixed Issues

$ https://github.com/Expensify/App/issues/23171
PROPOSAL: GH_LINK_ISSUE(COMMENT)

### Tests

Have two users A and B

### Active chat case

1. As a user A, open chat with user B and as user B open chat with user A in different tabs
2. Send a message as a user A to user B 
3. Ensure no marker is displayed to user A and user B

### Inactive chat case

1. As a user A, open chat with user B and as user B open chat other than user A in different tabs
2. Send a message as a user A to user B 
3. Ensure no marker is displayed to user A but user B sees new marker above new messages

### Mark as unread

1. As a user B, mark any message as unread from user A
2. Verify new marker appears above it
3. Go to any other chat and come back to chat with user A
4. Verify marker still appears
5. Go to any other chat and come back to chat with user A
6. Verify marker is not visible now

### Deleting a message

1. As a user A, open chat with user B and as user B open chat other than user A in different tabs
2. Send two message as a user A to user B 
3. Ensure no marker is displayed to user A but user B sees new marker above new messages
4. As a user A, delete message
5. As user B, ensure new marker line is there and message is deleted
6. As a user A, delete another message
7. As user B, ensure new marker line and message are deleted

### Scrolled up

1. As a user A, open chat with user B and as user B open chat other than user A in different tabs
2. Send a message as a user A to user B 
3. Ensure no marker is displayed to user A but user B sees new marker above new messages
4. Scroll up and ensure you see a floating button
5. Click on it and ensure you're taken down

- [x] Verify that no errors appear in the JS console

### Offline tests

None, same as online(As message sending and receiving is halt when offline)
### QA Steps

Please see Tests
- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/1676818/d568e0e9-6175-4250-94f5-22dfbf6871fd

https://github.com/Expensify/App/assets/1676818/dd7b4858-db4d-4bf2-94a9-5bc944236fc4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/1676818/20eac38d-52a4-46fd-9d6c-bdcbc87e9b44

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/1676818/d0831d97-ae2f-4e10-b6f4-16b12f87f1ba

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/1676818/90e5e812-8695-4789-92b0-bdae9ba75d13

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/1676818/da30b1fc-5eb3-4419-b528-a2c49265ed1c

</details>

### CODE REVIEW

File: src/libs/actions/IOU.js
Comment (MonilBhavsar) [inline src/libs/actions/IOU.js:1215]: @gedu any idea why this related to MoneyRequest was added here 😅

File: src/libs/actions/IOU.js
Comment (MonilBhavsar) [inline src/libs/actions/IOU.js:1259]: Same 😅

File: src/pages/home/report/ReportActionsList.js
Comment (gedu) [reply src/pages/home/report/ReportActionsList.js]: In the last commit, yes, it will be

File: src/pages/home/report/ReportActionsList.js
Comment (gedu) [reply src/pages/home/report/ReportActionsList.js:116]: We need to fire a rerender, there was a race condition the unread marker wasn't getting re-calculated because the rerender produced by onyx (when reports get updated) ran before the clearing `currentUnreadMarker.current = null;` so we couldn't use the onyx rerender for recalculations

**Review by allroundexperts — CHANGES_REQUESTED** [review]

File: src/pages/home/report/ReportActionsList.js
Comment (gedu) [reply src/pages/home/report/ReportActionsList.js:140]: We run into some weird issues trying to use `useRef` that was getting initialized cross-rerenders (when it shouldn't), couldn't find the root cause, will require more time, but seems it is related to Onyx and Navigation. So we have to move out the `prevReportID` to keep track of the last report id. Seems that prev refactor they [do the same](https://github.com/Expensify/App/pull/23729/files)
I need to check the `userActiveSince.current` isn't null to clean it, because Onyx sends the old report ID for a couple of frames, without that check the `useActiveSince` will be updated when it shouldn't

File: src/pages/home/report/ReportActionsList.js
Comment (gedu) [reply src/pages/home/report/ReportActionsList.js:120]: Just gave it around of testing using ref and is working, so I will remove the `useState(sortedReportActions.length);`

File: src/pages/home/report/ReportActionsList.js
Comment (gedu) [reply src/pages/home/report/ReportActionsList.js:1]: It is doing less than before, most of the logic where at `ReportActionsView`, but now are just 3 `useEffects`, also I'm not seeing a good way to extract the logic and put it into a hook

Comment (MonilBhavsar) [reply src/pages/home/report/ReportActionsList.js:1]: @allroundexperts thoughts on this^ comment please

Comment (MonilBhavsar) [reply src/pages/home/report/ReportActionsList.js:1]: Thank you! We're looking to merge this PR soon, so if you could please prioritise it

**Comment by gedu** [comment]
Yes, on my way

**Comment by gedu** [comment]
@allroundexperts I think all should be good

**Comment by MonilBhavsar** [comment]
> BUG / Inconsistency:
The message does not get unread after a tab is focused / refocused.

@allroundexperts could you please tell the steps here. Sorry, not getting it from video what exactly is happening there

**Comment by MonilBhavsar** [comment]
> chat auto scrolls to the bottom a little.

Good catch! Looks like it didn't. I think it's still above the threshold to view Floating button

**Comment by MonilBhavsar** [comment]
I'll add QA and tests steps, as there are also some backend changes that are on hold for this PR 
cc @allroundexperts @gedu

**Comment by gedu** [comment]
> **BUG / Inconsistency**: The message does not get unread after a tab is focused / refocused.
> 

I think is not a bug from the green unread marker, it behaves the same as you leave the chat an enters again. It keeps the green marker until you leave again.

So this behavior is good 
1. marking messages as unread
2. change tabs or report
3. going back and seeing the green marker
4. change tabs or report
5. going back and not seeing the green marker 

The bugs come from the LHN, the text isn't bold. @allroundexperts This is a bug already present on `main`

**Comment by gedu** [comment]
> BUG: On Android, if you scroll to the top of the chat and send a message from another account, on scrolling back to the

Taking a look at it, it should mark it yes, good catch

**Comment by gedu** [comment]
> BUG: On Android, if you scroll to the top of the chat and send a message from another account, on scrolling back to the

@allroundexperts  on Android I couldn't reproduce it, I updated this branch with main before testing, maybe there were some changes?


https://github.com/Expensify/App/assets/1676818/dd8826a7-82d1-4ebf-afd4-eb3f3ddcdba0

**Comment by MonilBhavsar** [comment]
Conflicts resolved, cleaned up, Tests and QA section is also updated

**Review by MonilBhavsar — APPROVED** [review]

**Comment by MonilBhavsar** [comment]
@allroundexperts all ready for review! Could you please take a look today

**Comment by MonilBhavsar** [comment]
Is it reproducible on main too?
As mentioned by @gedu this could be the cause https://github.com/Expensify/App/pull/18637#issuecomment-1675050658
And we can fix it in a separate PR

**Comment by MonilBhavsar** [comment]
I'll verify, please wait. Meanwhile could you please try having emulator in the same frame, so is not focussed again

**Comment by gedu** [comment]
> In my current tests, the InActive chat case that was defined in test section is failing.

Still trying to reproduce it, but on my side is working

https://github.com/Expensify/App/assets/1676818/aab8fbc8-9613-40b5-9db2-73f7d549f1c6

**Comment by MonilBhavsar** [comment]
@allroundexperts that video from @gedu is from Android native, no? 😅 
Is it now working on your side?

**Comment by MonilBhavsar** [comment]
I don't think so, as that PR is not merged yet

**Review by allroundexperts — APPROVED** [review]
This tests well from my side. Thanks @MonilBhavsar and @gedu for your patience!

**Comment by MonilBhavsar** [comment]
Thanks! I can verify on my side

**Review by allroundexperts — APPROVED** [review]

**Comment by MonilBhavsar** [comment]
Confirmed! works for me too
https://drive.google.com/file/d/1HnXkqPt5VKGt0UB2A6dOPnd3UByjDPIf/view?usp=drive_link

**Comment by Pujan92** [comment]
@gedu I think we should unsubscribe the typing event in the `ReportActionsView` on unmount instead of keeping it open.
https://github.com/Expensify/App/pull/18637/files#diff-3d3d60df30d1761a83ff8413b06ecef63fcd0084053cd70d5686098e71d87362L184

Need to create cleanup function here which calls `Report.unsubscribeFromReportChannel(reportID);`
https://github.com/Expensify/App/blob/c435734d8cf360e964e20eea1fb54c541f6fa7ea/src/pages/home/report/ReportActionsView.js#L86-L89

**Comment by gedu** [comment]
> @gedu I think we should unsubscribe the typing event in the `ReportActionsView` on unmount instead of keeping it open or restricting multiple subscriptions for the same channel. https://github.com/Expensify/App/pull/18637/files#diff-3d3d60df30d1761a83ff8413b06ecef63fcd0084053cd70d5686098e71d87362L184
> 
> Need to create cleanup function here which calls `Report.unsubscribeFromReportChannel(reportID);`
> 
> https://github.com/Expensify/App/blob/c435734d8cf360e964e20eea1fb54c541f6fa7ea/src/pages/home/report/ReportActionsView.js#L86-L89

Hey, the next `useEffect` runs the `unsubscribeFromReportChannel`. 
All those flows, weren't part of unread marker refactor, come from so many merges

**Comment by hayata-suenaga** [comment]
@gedu could you check if [this issue](https://github.com/Expensify/App/issues/25606) was caused by this PR?

**Comment by roryabraham** [comment]
Evaluating the deploy blockers related to this PR, it seems there are three:

- https://github.com/Expensify/App/issues/25776
- https://github.com/Expensify/App/issues/25673
- https://github.com/Expensify/App/issues/25737

File: src/pages/home/report/ReportActionsView.js
Comment (mollfpr) [inline src/pages/home/report/ReportActionsView.js:155]: Hi, coming from https://github.com/Expensify/App/issues/26432

The issue is a regression from this removed, where the current user doesn't scroll to the bottom after sending a message.

I have several questions regarding these changes:
1. Are the above results expected?
2. If not, can we reintegrate the listener `Report.subscribeToNewActionEvent` to scroll the current user when sending a message? Or there's a performance consideration to use that listener.

cc @gedu @MonilBhavsar @allroundexperts

Comment (MonilBhavsar) [reply src/pages/home/report/ReportActionsView.js:155]: > Are the above results expected?

Sorry, what result?

Comment (mollfpr) [reply src/pages/home/report/ReportActionsView.js:155]: Ahh, sorry for not making it clear.

> Sorry, what result?

The current user doesn't scroll to the bottom after sending a message when they're not at the bottom of the chat.

Comment (MonilBhavsar) [reply src/pages/home/report/ReportActionsView.js:155]: That does seem like issue. We should definitely scroll user to the bottom

Comment (mollfpr) [reply src/pages/home/report/ReportActionsView.js:155]: @MonilBhavsar Is there any drawback or performance issue if we use `Report.subscribeToNewActionEvent` to get the current report action?

Comment (gedu) [reply src/pages/home/report/ReportActionsView.js:155]: I don't think it will have an impact on performance, it will be mainly just to know when a new message is sent. Because this newActionEvent only listens to comments sent, and not received from others.

File: src/pages/home/report/ReportActionsList.js
Comment (alitoshmatov) [inline src/pages/home/report/ReportActionsList.js:322]: `keyboardShouldPersistTaps` is set to "handled" is causing keyboard not closing when tapped outside. https://github.com/Expensify/App/issues/53196

**Pull request merged**
